papers/leanimt/paper/leanimt-paper.tex

```diff
@@ -0,0 +1,1497 @@
+                        $H_2$ [$H_0$ [$a_0$] [$a_1$]] [$H_1$ [$a_2$] [$a_3$]]
+                    ]
+                    [$a_4$ [$a_4$ [$a_4$]]]
+                ]
+            \end{forest}
+        \end{center}
+        \begin{center}
+            \textit{Before inserting $a_5$}
+        \end{center}
+        \begin{center}
+            \begin{forest}
+                for tree={edge path={\noexpand\path[\forestoption{edge}] (\forestOve{\forestove{@parent}}{name}.parent anchor) -- +(0,-12pt)-| (\forestove{name}.child anchor)\forestoption{edge label};}, rectangle, draw, align=center}
+                [$H_5$, draw=blue600, top color=blue600!0, bottom color=blue600!0, line width=1pt
+                [
+                $H_2$ [$H_0$ [$a_0$] [$a_1$]] [$H_1$ [$a_2$] [$a_3$]]
+                ]
+                [$H_4$ , edge={blue600, line width=1pt}, draw=blue600, top color=blue600!0, bottom color=blue600!0, line width=1pt [$H_4$ , edge={blue600, line width=1pt}, draw=blue600, top color=blue600!0, bottom color=blue600!0, line width=1pt [$a_4$] [$a_5$ , edge={blue600, line width=1pt}, draw=blue600, top color=blue600!0, bottom color=blue600!0, line width=1pt]]]
+                ]
+            \end{forest}
+        \end{center}
+        \begin{center}
+            \textit{After inserting $a_5$}
+        \end{center}
+    \end{multicols}
+}
+
+
+\subsubsection{Pseudocode}
+
+\begin{algorithm}[H]
+    \caption{LeanIMT Insert algorithm}\label{insert}
+    \begin{algorithmic}[1]
+        \Procedure{Insert}{$leaf$}
+        \If{$depth < newDepth$} \Comment{$newDepth$ is the new depth of the tree after inserting the new node}
+        \State add a new empty array to nodes \Comment{Add a new tree level}
+        \EndIf
+        \State $node\gets leaf$
+        \State $index\gets size$ \Comment{The index of the new leaf equals the number of leaves in the tree.}
+        \For{level from 0 to depth - 1}
+        \State nodes[level][index] $\gets$ node
+        \If{index \textbf{is odd}} \Comment{It's a right node}
+        \State sibling $\gets$ nodes[level][index - 1]
+        \State node $\gets$ \textbf{hash}(sibling, node)
+        \EndIf
+        \State index $\gets$ $\lfloor index/2 \rfloor$ \Comment{Divides the index by 2 and discards the remainder.}
+        \EndFor
+        \State nodes[depth] $\gets$ [node] \Comment{Store the new root at the top level}
+        \EndProcedure
+    \end{algorithmic}
+\end{algorithm}
+
+
+\subsubsection{Time complexity}
+
+$n$: Number of leaves in the tree.
+
+$d$: Tree depth after inserting the new leaf.
+
+Every time a new node is added, it is necessary to update or add the ancestors up to the root of the tree.
+
+Number of operations when adding a leaf: $d + 1$
+
+$d + 1 = O(\log n) + 1$
+
+$\phantom{d + 1} \Rightarrow \boxed{O(\log n)}$
+
+$d = \lceil \log (n+1) \rceil \leq \log (n+1) + 1$
+
+$\phantom{d = \lceil \log (n+1) \rceil} \leq O(\log n) + 1$
+
+$\phantom{d = \lceil \log (n+1) \rceil} \Rightarrow O(\log n)$
```

Comment:
nit: 
my former math teacher would be mad at this ^^. Because it abuses notation a lot and isn't strictly rigorous.
Maybe the first line is a bit redundant too?

Isn't it enough to write something like:
Given that `d = ceil(log(n+1))` for a **balanced** merkle tree, we can write
`d = ceil(log(n+1)) <= log(n+1) + 1`.
and because `log(n+1) + 1 ∈ O(log n)`,
therefore the time complexity of `log n`

Reply:
Hey! I think that's fine and it's also fine to say `O(log n)` directly. But I tried to write step by step so that people can understand the process.